crumbles/src/lib.rs

```diff
@@ -0,0 +1,653 @@
+/// track of which pages have been written to. Each page is [`PAGE_SIZE`] in
+/// size.
+///
+/// A `Mmap` may be backed by a set of files, physical memory, or a combination
+/// of both. Use [`new`] to create a new mmap backed entirely by physical
+/// memory, and [`with_files`] to create a new mmap backed by the given
+/// set of files, at the given offsets.
+///
+/// It is possible to create snapshots of the memory, which can be used to
+/// revert to a previous state. See [`snap`] for more details.
+///
+/// Writes are tracked at the page level. This functions as follows:
+///
+/// - When a region is first mapped, its permissions are set to read-only,
+///   resulting in a segmentation fault when a write is attempted.
+/// - When a write is attempted the segmentation fault is caught using a signal
+///   handler, which proceeds to set the permissions of the page to read-write
+///   while also marking it as dirty in the mmap.
+///
+/// `Mmap` is [`Sync`] and [`Send`].
+///
+/// [`new`]: Mmap::new
+/// [`with_files`]: Mmap::with_files
+/// [`snap`]: Mmap::snap
+#[derive(Debug)]
+pub struct Mmap(&'static mut MmapInner);
+
+impl Mmap {
+    /// Create a new mmap, backed entirely by physical memory. The memory is
+    /// initialized to all zeros.
+    ///
+    /// # Errors
+    /// If the underlying call to map memory fails, the function will return an
+    /// error.
+    ///
+    /// # Example
+    /// ```rust
+    /// use crumbles::Mmap;
+    ///
+    /// let mmap = Mmap::new().unwrap();
+    /// assert_eq!(mmap[..0x10_000], [0; 0x10_000]);
+    /// ```
+    pub fn new() -> io::Result<Self> {
+        unsafe { Self::with_files(None) }
+    }
+
+    /// Create a new memory, backed by the given files at the given offsets.
+    /// Regions of the memory not covered
+    ///
+    /// # Errors
+    /// If the given files are too large for the [`MEM_SIZE`] memory region, or
+    /// at an offset where they wouldn't fit within said region, the function
+    /// will return an error. Also, if any of the files' size or offsets is not
+    /// a multiple of the page size, the function will return an error.
+    ///
+    /// # Safety
+    /// The caller must ensure that the given files are not modified while
+    /// they're mapped. Modifying a file while it's mapped will result in
+    /// *Undefined Behavior* (UB).
+    ///
+    /// # Example
+    /// ```rust
+    /// use std::fs::File;
+    /// use std::io::Read;
+    /// use std::iter;
+    ///
+    /// use crumbles::Mmap;
+    ///
+    /// let mut file = File::open("LICENSE").unwrap();
+    ///
+    /// let mut contents = Vec::new();
+    /// file.read_to_end(&mut contents).unwrap();
+    ///
+    /// let mmap = unsafe { Mmap::with_files(iter::once(Ok((file, 0)))) }.unwrap();
+    /// assert_eq!(mmap[..contents.len()], contents[..]);
+    /// ```
+    pub unsafe fn with_files<I>(files_and_offsets: I) -> io::Result<Self>
+    where
+        I: IntoIterator<Item = io::Result<(File, usize)>>,
```

Comment:
As a slightly pedantic suggestion I would prefer to have the iterator iterate over `(offset, file)` and not `(file, offset)` purely because this is the tuple that is also returned by a call to `enumerate()`.
But feel free to ignore this

Reply:
Ah, interesting. There would be no need to `.map` and switch the order. Will do.